### PR TITLE
MCCAS: Implement Tag-Splitting strategy for debug info

### PR DESCRIPTION
**Pull request by felipepiovezan**
Each commit is intended to be reviewed independently for ease of review.
Please read each commit message.

### CODE REVIEW

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.def
Comment (rastogishubham) [inline llvm/include/llvm/MC/CAS/MCCASObjectV1.def:40]: Just to be sure, the whole point of this is to maintain the order of the abbrevs again for materialization right?

File: llvm/include/llvm/MC/CAS/MCCASDebugV1.h
Comment (rastogishubham) [inline llvm/include/llvm/MC/CAS/MCCASDebugV1.h:59]: Nice!

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.def
Comment (rastogishubham) [inline llvm/include/llvm/MC/CAS/MCCASObjectV1.def:41]: Is there one DIETopLevelRef per translation unit? If so, why is it any different from a`DebugInfoSectionRef`?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp:2891]: To future-proof this, why don't we make this function a member of `DIEToCASConverter` and add all the tags we care about into a denseset?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp:2896]: A comment on why you are splitting on DW_AT_low_pc would be nice!

**Review by rastogishubham — COMMENTED** [review]
Looks good, some comments inline

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.def
Comment (felipepiovezan) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.def:40]: Not quite! The point here is so that we can replace CAS references with indices into the the list of references of this block, thus saving 7 bytes per reference

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.def
Comment (felipepiovezan) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.def:41]: >Is there one DIETopLevelRef per translation unit? 

Yeah

> If so, why is it any different from a DebugInfoSectionRef?

This is both to ease the transition period and also to put all blocks semantically related under the same parent, so that they are easy to find and materialize. The cost is negligible.

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp:2891]: A member function indicates that it requires knowledge of the implementation of the class. In this case, `shouldCreateSeparateBlockFor ` needs no private members of `DIEToCASConverter` (in fact, it is completely independent of the class), as such it doesn't need to be a member.

>  add all the tags we care about into a denseset?

The problem with this is that the queries aren't as simple as "is this tag in a blessed list?".
More likely, each tag will have its own heuristic, like the low_pc one below

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp:2896]: Yup, good point!

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp:2941]: This is minor, but all the code that does iteration over DIEs uses while loops, maybe this could too? Just to be consistent

File: llvm/test/tools/llvm-cas-dump/basic_debug_test.ll
Comment (rastogishubham) [inline llvm/test/tools/llvm-cas-dump/basic_debug_test.ll]: Nice!

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.def
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.def:40]: I see, that makes sense!

File: llvm/include/llvm/MC/CAS/MCCASObjectV1.def
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASObjectV1.def:41]: Sounds good to me

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp:2891]: That is fair

**Review by rastogishubham — APPROVED** [review]
Once you fix the merge conflict, I think it looks good!

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp:2941]: There is some semantic meaning in the choice here that would get lost. 
In general, a while loop is telling us that we don't want / can't update the induction variable every iteration, or that there is no clear induction variable.
A for loop is communicating that intent very clearly: `Child` is the induction variable, and we  want to update it on every iteration. Note how, if we replace this with a `while`, we would have to be careful to add a `Child = Child.getSibling()` in more than one location in the loop body. Furthermore, future code changes could be buggy as a result of the developer having to remember to manually update the induction variable.

**Comment by felipepiovezan** [comment]
@swift-ci test

**Comment by felipepiovezan** [comment]
@swift-ci test

**Comment by felipepiovezan** [comment]
@swift-ci test

**Comment by felipepiovezan** [comment]
Rebased

**Comment by felipepiovezan** [comment]
@swift-ci test

**Comment by felipepiovezan** [comment]
rebased

**Comment by felipepiovezan** [comment]
@swift-ci test

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp:3181]: Should we assume the endianness to be little always? Isn't it stored in the MCObjectProxy?

`  if (HeaderRef::Cast(MCObj)) {
    MachOHeaderParser P;
    if (Error Err = P.parse(MCObj.getData()))
      return Err;
    Is64Bit = P.Is64Bit;
    IsLittleEndian = P.IsLittleEndian;`

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Can we rename this to something like `startDIERef`? Just curious if you think that makes more sense

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp:3049]: The same here, should we assume little endian?

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp:3050]: Also the AddrSize, I am not sure if these will ever change, but maybe these shouldn't be constexprs

**Review by rastogishubham — APPROVED** [review]
Overall LGTM, minor tweaks if you think they should be addressed

**Comment by felipepiovezan** [comment]
To record what we discussed on slack: I intend to change those assumptions, but we need an easier mechanism to access this information from the MachO header block. It will be easier to do this once we remove the older implementation

File: llvm/lib/MC/MCCASObjectV1.cpp
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp]: Good idea, will do!

**Comment by felipepiovezan** [comment]
rebased, addressed review comment

**Comment by felipepiovezan** [comment]
@swift-ci test

**Pull request merged**
